src/utils.rs

```diff
@@ -99,6 +98,42 @@ impl HRESULT {
     }
 }
 
+/// Wraps the most common result combinations from [`DxcOperationResult`].  Create with
+/// [`Self::from_operation_result()`].
+#[derive(Clone, Debug)]
+pub enum OperationResult {
+    /// Operation succeeded
+    Success(Vec<u8>),
+    /// Operation failed
+    Failed(String),
+    /// Operation succeeded with warnings
+    SuccessWithWarnings(String, Vec<u8>),
```

Comment:
```suggestion
    Success { warnings: Option<String>, blob: Vec<u8>},
    /// Operation failed
    Failed(String),
```

Wouldn't it be much more natural to express like this? That way, if there are errors we can just ignore them if we want (or not), but at least we can't forget to process successful results (which is the possibility now - one can forget to process `SuccessWIthWarnings` in a match statement pretty easily).

Reply:
Probably that field should be called something other then `warnings` since it'll most likely include more of the compiler output?

Reply:
It seems pretty hard to miss `SuccessWithWarnings` unless someone _intentionally_ doesn't look at the variants and maps whatever is _not_ `Success(x)` to disappear in the void (in a `match` or `if let Success(x)` block): in most cases folks place a `panic!()` here or bubble it up in a `Result` of sorts so they should catch it quickly enough if actually testing their code (and having warnings somewhere)...

I've contemplated quite a few designs for the return type and while your suggestion is definitely one that I've considered because it makes the handling easier:

```rust
match compile() {
    Success { warnings, blob } => {
        if let Some(warnings) = warnings {
            eprintln / warn!
        }
        blob
    }
    ...
}
````

This however turns it right back into a 2-variant enum, that would be more convenient to reuse and represent as a `Result<CompileSuccess, CompileError>` where `CompileSuccess` contains a `warnings` and `blob` field. But then it's very easy to miss the warnings:

```rust
let blob = compile()?.blob;
```

:sweat_smile: 

(And `CompileError` here would be some combination of the `Failed()` arm and the original `HassleError` low-level library errors)

---

I've also been thinking about a `CompileResult { warnings: Option<String>, blob: Option<Vec<u8>> }` but that has a funny fourth `None,None` state 
(And note that we still need to result the lower-level `HassleError` enum somewhere else).

Reply:
> that would be more convenient to reuse and represent as a `Result<CompileSuccess, CompileError>`

Sort of, but you give a good reason not to turn this into `Result` ;-)

Reply:
With the suggested `enum` above, it's harder to miss `.messages` than the struct because it must be matched to get the blob out.

Following that I've just been contemplating whether to return a `Result<OperationResult>` _or_ change `OperationResult::Failed` to contain a `HassleError` and/or `String` variant (or a third variant). The latter two disallow `?`'ing internally unless we have an `Into` variant.

Reply:
Doing this makes it impossible to use the `?` operator in most places in `hassle-rs`, for message-less `HRESULT`s out of other APIs (before checking the compile status and actioning on that). Contemplating to keep the variant that makes it easier to "miss" `.warnings`/`.messages` over heavily complicating the code...

Reply:
If anything we could merge this now, and decide on improving it even further later on if it's really too easy to miss the warnings and/or someone finds a better solution in the future.